File: Checklists_Code/Checklists_Code/AllListsViewController.m
//
//  UITableViewController.m
//  Checklists_Code
//
//  Created by wangx on 15/6/24.
//  Copyright (c) 2015年 wxiang1991. All rights reserved.
//

#import "Checklist.h"
#import "AllListsViewController.h"
#import "ListDetailViewController.h"
#import "ChecklistViewController.h"

@implementation AllListsViewController

-(void)loadView{
    [super loadView];
    
    
    UIBarButtonItem *addBarButtonItem=[[UIBarButtonItem alloc]initWithBarButtonSystemItem:UIBarButtonSystemItemAdd target:self action:@selector(AddChecklist)];
    self.navigationItem.rightBarButtonItem=addBarButtonItem;
}
- (void)viewDidLoad {
    [super viewDidLoad];
//    self.lists=[[NSMutableArray alloc]initWithCapacity:20];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return [self.dataModel.lists count];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    static NSString *CellIdentifer=@"Cell";
        UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifer];
    if (cell==nil) {
        cell=[[UITableViewCell alloc]initWithStyle:UITableViewCellStyleSubtitle reuseIdentifier:CellIdentifer];
    }
    Checklist *checklist=self.dataModel.lists[indexPath.row];
    cell.textLabel.text=checklist.name;
    cell.imageView.image=[UIImage imageNamed:checklist.iconName];
    cell.accessoryType=UITableViewCellAccessoryDetailDisclosureButton;
    int count=[checklist countUncheckedItems];
    if([checklist.items count]==0){
        cell.detailTextLabel.text = @"(No Items)";
    }else if(count ==0){
        cell.detailTextLabel.text =@"全部搞定收工！";
    }else{
        cell.detailTextLabel.text = [NSString stringWithFormat:@"%d Remaining",count];
    }
    
    cell.imageView.image=[UIImage imageNamed:checklist.iconName];
    return cell;
}

//点击加号进入
-(void)AddChecklist{
    ListDetailViewController *listDetailVC=[[ListDetailViewController alloc]initWithStyle:UITableViewStyleGrouped];
    listDetailVC.delegate=self;
    [self.navigationController pushViewController:listDetailVC animated:YES];
}
//点击tableviewcell中的小详细按钮进行描述方法中
- (void)tableView:(UITableView *)tableView accessoryButtonTappedForRowWithIndexPath:(NSIndexPath *)indexPath{
    ListDetailViewController *controller = [[ListDetailViewController alloc]initWithStyle:UITableViewStyleGrouped];
    controller.delegate = self;
    Checklist *checklist=self.dataModel.lists[indexPath.row];
    controller.checklistToEdit=checklist;
    [self.navigationController pushViewController:controller animated:YES];
}
//是否可以删除
- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath{
    return YES;
}
//删除是触发的方法
-(void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath{
    if (editingStyle==UITableViewCellEditingStyleDelete) {
        //删除list中的内容
        [self.dataModel.lists removeObjectAtIndex:indexPath.row];
        //在tableview中删除该列
        [tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationAutomatic];
    }
}

-(void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath{
    ChecklistViewController *controller=[[ChecklistViewController alloc]init];
    Checklist *checklist=self.dataModel.lists[indexPath.row];
    controller.checklist=checklist;
    controller.delegate=self;
    [self.navigationController pushViewController:controller animated:YES];
}

//实现ListDetailViewControllerDelegate中的三个方法
- (void)listDetailViewControllerDidCancel:(ListDetailViewController *)controller
{
    [self.navigationController popToViewController:self animated:YES];
}

- (void)listDetailViewController:(ListDetailViewController *)controller didFinishAddingChecklist:(Checklist *)checklist
{
    [self.dataModel.lists addObject:checklist];
    [self.dataModel sortChecklists];
    [self.tableView reloadData];
    [self.navigationController popToViewController:self animated:YES];
}

- (void)listDetailViewController:(ListDetailViewController *)controller didFinishEditingChecklist:(Checklist *)checklist
{
    [self.dataModel sortChecklists];
    [self.tableView reloadData];
    [self.navigationController popToViewController:self animated:YES];
}

-(void)ChecklistViewControllerDidBack:(ChecklistViewController *)controller{
    [self.tableView reloadData];
    [self.navigationController popToViewController:self animated:YES];
}
@end

File: ChecklistSwift/ChecklistSwift/AllListsViewController.swift
//
//  AllListsViewController.swift
//  ChecklistSwift
//
//  Created by 王享 on 15/12/25.
//  Copyright © 2015年 王享. All rights reserved.
//

import UIKit

class AllListsViewController: UITableViewController,ListDetailViewControllerDelegate,ChecklistViewControllerDelegate{

    var dataModel:DataModel?;
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.initSubView();
    }
    
    func initSubView(){
        let addBarButtonItem:UIBarButtonItem = UIBarButtonItem(barButtonSystemItem:UIBarButtonSystemItem.add, target: self, action: #selector(AllListsViewController.AddChecklist))
        self.navigationItem.rightBarButtonItem = addBarButtonItem
    }
    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1;
    }
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        if(self.dataModel!.lists == nil){
            return 0
        }else{
            return self.dataModel!.lists!.count
        }
    }

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let identifer:String? = "Cell"
        var cell:UITableViewCell? = tableView.dequeueReusableCell(withIdentifier: identifer!)
        if cell == nil {
            cell = UITableViewCell(style: UITableViewCellStyle.subtitle, reuseIdentifier:identifer);
        }
        cell!.textLabel?.text="测试"
        let checklist:Checklist = self.dataModel!.lists![indexPath.row]
        cell!.textLabel?.text=checklist.name
        if checklist.iconName != nil {
            cell!.imageView?.image=UIImage(named: checklist.iconName!)
        }
        cell!.accessoryType=UITableViewCellAccessoryType.detailDisclosureButton
        let count:Int = checklist.countUncheckedItems()
        if checklist.items != nil {
            if checklist.items!.count == 0 {
                cell!.detailTextLabel?.text = "(No Items)"
            }
        }
        if count == 0 {
            cell!.detailTextLabel?.text = "全部搞定收工"
        }else{
            cell!.detailTextLabel?.text = "".appendingFormat("%d Remaining", count)
        }
        return cell!;
    }
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let controller:ChecklistViewController = ChecklistViewController()
        let checklist:Checklist = self.dataModel!.lists![indexPath.row]
        controller.checklist = checklist
        controller.delegate=self
        self.navigationController?.pushViewController(controller, animated: true)
    }
    
    override func tableView(_ tableView: UITableView, accessoryButtonTappedForRowWith indexPath: IndexPath) {
        let controller:ListDetailViewController = ListDetailViewController.init(style:UITableViewStyle.grouped);
        controller.delegate=self
        let checklist:Checklist = self.dataModel!.lists![indexPath.row]
        controller.checklistToEdit=checklist
        self.navigationController?.pushViewController(controller, animated: true)
    }
    
    override func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
        return true
    }
    
    override func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCellEditingStyle, forRowAt indexPath: IndexPath) {
        if editingStyle == UITableViewCellEditingStyle.delete {
            self.dataModel!.lists!.remove(at: indexPath.row)
            tableView.deleteRows(at: [indexPath], with: UITableViewRowAnimation.automatic)
        }
    }
    
    func AddChecklist(){
        let controller:ListDetailViewController = ListDetailViewController.init(style:UITableViewStyle.grouped);
        controller.delegate=self
        self.navigationController?.pushViewController(controller, animated: true)
    }
    
    func ChecklistViewControllerDidBack(_ controller:ChecklistViewController){
        self.tableView.reloadData()
        self.navigationController?.popToViewController(self, animated: true)
    }
    
    func listDetailViewControllerDidCancel(_ controller:ListDetailViewController){
        self.navigationController?.popToViewController(self, animated: true)
    }
    
    func listDetailViewController(_ controller:ListDetailViewController,didFinishAddingChecklist checklist:Checklist){
        if(self.dataModel!.lists == nil){
            self.dataModel!.lists = Array<Checklist>();
        }
        self.dataModel?.lists?.append(checklist)
        
        self.tableView.reloadData()
        self.navigationController?.popToViewController(self, animated: true)
    }
    
    func listDetailViewController(_ controller:ListDetailViewController,didFinishEditingChecklist checklist:Checklist){
        self.tableView.reloadData()
        self.navigationController?.popToViewController(self, animated: true)
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }

}
